crates/datasources/src/excel/table.rs

```diff
@@ -0,0 +1,64 @@
+use std::borrow::Cow;
+use std::sync::Arc;
+
+use calamine::{open_workbook, Reader, Xlsx};
+use datafusion::datasource::streaming::StreamingTable;
+use datafusion::datasource::TableProvider;
+use datafusion::physical_plan::streaming::PartitionStream;
+
+use super::xlsx_sheet_value_to_record_batch;
+use crate::common::url::DatasourceUrl;
+use crate::excel;
+use crate::excel::errors::ExcelError;
+use crate::excel::stream::ExcelSheetPartitionStream;
+use crate::object_store::generic::GenericStoreAccess;
+
+pub async fn excel_streaming_table(
+    _store_access: GenericStoreAccess,
+    source_url: DatasourceUrl,
+    sheet_name: Option<&str>,
+    has_header: bool,
+    infer_schema_length: usize,
+) -> Result<Arc<dyn TableProvider>, ExcelError> {
+    let mut workbook: Xlsx<_> = match source_url {
+        DatasourceUrl::File(path) => open_workbook(path)?,
+        DatasourceUrl::Url(_) => todo!("remote files not supported yet"),
+    };
+
+    // for now, this only infers table data from only the first sheet in a workbook
+    // when read_excel gets vectorised to real columns this should be updated
+    let sheet = sheet_name.map(Cow::Borrowed).unwrap_or_else(|| {
+        let sheets = workbook.sheet_names();
+        let first = sheets.first().expect("sheet is not empty");
+        Cow::Owned(first.clone())
+    });
+
+    let working_sheet_range = match workbook.worksheet_range(&sheet) {
+        Ok(r) => Ok(r),
+        Err(_) => Err(ExcelError::Load {
+            msg: "Failed to open .xlsx file.".to_owned(),
+        }),
+    };
+
+    let range = working_sheet_range?;
+    let schema = excel::infer_schema(&range, has_header, infer_schema_length)?;
+
+    // todo: expand this handle multiple sheets
```

Comment:
capitalize TODO; 

I think there's a lot of possible options here for dealing with multiple sheets, but the one that sort of makes the most sense to me is to treat sheets as tables and workbooks as databases. So we might have a different implementation that would handle that.

Reply:
yupI was thinking this too! will likely follow this up with this bit in separate pr